tools/please_pex/pex_main.py

```diff
@@ -141,27 +143,65 @@ def get_code(self, fullname):
         return module.__loader__.get_code(fullname)
 
 
+def pex_basepath(temp=False):
+    if temp:
+        import tempfile
+        return tempfile.mkdtemp(dir=os.environ.get('TEMP_DIR'), prefix='pex_')
+    else:
+        return os.path.expanduser('~/.pex')
+
+def pex_uniquedir():
+    return f'pex-{PEX_STAMP}'
+
+
 def explode_zip():
     """Extracts the current pex to a temp directory where we can import everything from.
 
     This is primarily used for binary extensions which can't be imported directly from
     inside a zipfile.
     """
-    import contextlib, shutil, tempfile, zipfile
+    import contextlib
+
+    @contextlib.contextmanager
+    def pex_lockfile(basepath, uniquedir):
+        # Acquire the lockfile.
+        lockfile_path = os.path.join(basepath, f'.lock-{uniquedir}')
```

Comment:
do you need the lockfile to be different to the directory? wondering whether you could lock on the directory instead of the file and use its presence instead of writing to it?

Doesn't really matter a huge amount, just curious. I'm not 100% sure you can `flock()` a directory...

Reply:
Discussed offline in [gitter](https://gitter.im/please-build/Lobby?at=5d128b9a527a392800c1103e).

> I wanted to keep a separate state-bit for "This directory is ready for use", which lives outside the directory itself.